lib/Sema/ConstraintSystem.cpp

```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
   // Merge nodes in the constraint graph.
   CG.mergeNodes(typeVar1, typeVar2);
 
+  // Let's try to de-duplicate any generic requirements
+  // associated with given type variable, if it represents
+  // a generic parameter.
+  if (typeVar1->getImpl().getGenericParameter()) {
```

Comment:
Should we be looking at all type variables in the equivalence class of `typeVar1` to see if any of them have a corresponding generic parameter?

Reply:
I'm not actually sure what do you mean. I think the API doesn't have the best name here, I'll add `isGenericParameter` to make it clearer that this type variable is supposed to represent a particular generic parameter.

Reply:
We might have two type variables `$T1` and `$T2` that are always equivalent to each other, but only one of them has an associated generic parameter. If `typeVar1` ends up being `$T1` but `$T2` is the generic parameter, we would skip this code. Is that the right behavior?

Reply:
Interesting thought! As I understand current behavior, we only merge type variables representing generic parameters together as a result of `matchDeepEqualityTypes` so both sides have to be generic parameters. For argument/parameter conversions we wait until at least one side is resolved, same for `BindParam`. This check is not strictly necessary I just wanted to make sure that we don't call `gatherConstraints` after every merge. Maybe I should just do `getRepresentative` if check whether it's a generic parameter.

Reply:
I will also try to make this work cache of "fixed" requirements I have.

Reply:
Looks like de-duplication is not longer indeed with caching of fixed requirements, so this is none issue.